src/agent/rustjail/src/cgroups/fs/mod.rs

```diff
@@ -27,1120 +43,848 @@ macro_rules! sl {
-    if v != 0 {
-        write_file(dir, file, v.to_string())?;
-    }
+        if r.memory.is_some() {
+            info!(sl!(), "cgroup manager set memory !!!");
+            let mem_controller: &MemController = cg.controller_of().unwrap();
+            let memory = r.memory.as_ref().unwrap();
 
-    Ok(())
-}
+            if !update {
+                // initialize kmem limits for accounting
+                mem_controller.set_kmem_limit(1);
+                mem_controller.set_kmem_limit(-1);
+            }
 
-fn try_write_nonzero(dir: &str, file: &str, v: i128) -> Result<()> {
-    match write_nonzero(dir, file, v) {
-        Ok(_) => Ok(()),
-        Err(ref e) if io_error_kind_eq(e, std::io::ErrorKind::PermissionDenied) => Ok(()),
-        Err(e) => Err(anyhow!(e)),
-    }
-}
+            let limit = memory.limit.unwrap_or(0);
+            if limit != 0 {
+                mem_controller.set_limit(limit);
+            }
 
-fn remove(dir: &str) -> Result<()> {
-    fs::remove_dir_all(dir)?;
-    Ok(())
-}
+            let reservation = memory.reservation.unwrap_or(0);
+            if reservation != 0 {
+                mem_controller.set_soft_limit(reservation);
+            }
 
-fn apply(dir: &str, pid: pid_t) -> Result<()> {
-    write_file(dir, CGROUP_PROCS, pid)?;
-    Ok(())
-}
+            let swap = memory.swap.unwrap_or(0);
+            if cg.v2() {
+                let swap = convert_memory_swap_to_v2_value(swap, limit)?;
+            }
+            mem_controller.set_memswap_limit(swap);
 
-fn try_write_file<T: ToString>(dir: &str, file: &str, v: T) -> Result<()> {
-    match write_file(dir, file, v) {
-        Err(e) => {
-            let err = Errno::last();
-            if err == Errno::EINVAL || err == Errno::ENODEV || err == Errno::ERANGE {
-                warn!(sl!(), "Invalid Arguments!");
-                return Ok(());
+            let kernel = memory.kernel.unwrap_or(0);
+            if kernel != 0 {
+                mem_controller.set_kmem_limit(kernel);
             }
 
-            info!(sl!(), "{}", err.desc());
+            let kernel_tcp = memory.kernel_tcp.unwrap_or(0);
+            if kernel_tcp != 0 {
+                mem_controller.set_tcp_limit(kernel_tcp);
+            }
 
-            return Err(anyhow!(e));
+            if memory.swapiness.unwrap_or(0) <= 100 {
+                mem_controller.set_swappiness(memory.swapiness.unwrap_or(0) as u64);
+            }
+            if memory.disable_oom_killer.unwrap_or(false) {
+                match mem_controller.disable_oom_killer() {
+                    Ok(_) => {}
+                    Err(err) => {
+                        error!(sl!(), "failed to disable_oom_killer: {:?}", err);
+                    }
+                }
+            }
+            // res.memory.update_values = true;
         }
 
-        Ok(_) => {}
-    }
+        if r.pids.is_some() {
+            info!(sl!(), "cgroup manager set pids !!!");
```

Comment:
Another `!!!`

Reply:
This repeated use of `info!(sl!(), ...)` is another good candidate for a macro like `sl_info!`.